napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,150 @@
+from __future__ import annotations
+
+import logging
+from concurrent.futures import Executor, Future, ThreadPoolExecutor
+from threading import RLock
+from typing import Dict, Iterable, Optional, Tuple
+
+from napari.components import Dims
+from napari.layers import Layer
+from napari.utils.events.event import EmitterGroup, Event
+
+logger = logging.getLogger("napari.components._layer_slicer")
+
+
+class _LayerSlicer:
+    """
+    High level class to control the creation of a slice (via a slice request),
+    submit it (synchronously or asynchronously) to a thread pool, and emit the
+    results when complete.
+
+    Attributes
+    ----------
+    events : napari.utils.events.event.EmitterGroup
+        vispy event emitter
+    _executor : concurrent.futures.ThreadPoolExecutor
+        manager for the slicing threading
+    _layers_to_task : dict
+        task storage for cancellation logic
+    lock_layers_to_task : threading.RLock
+        lock to guard against changes to `_layers_to_task` when finding or
+        removing tasks.
```

Comment:
```suggestion
        lock to guard against changes to `_layers_to_task` when finding,
        adding, or removing tasks.
```

Reply:
I started to including `adding` as well. Theoretically it can be used for that as well, though we don't currently. Happy to add it to the doc string.